### PR TITLE
[SourceKit NameLookup] Don't walk into inactive clauses.

### PR DESCRIPTION
**Pull request by davidungar**
Don't do lookups in inactive #if clauses as long as ASTScopes do not support them.
Addresses rdar://56760957

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@nathawes This PR addresses the issue by not walking into inactive clauses. I realize that this fix may reduce the utility of SourceKit in these cases. Is this limitation acceptable? If not, I see two alternatives:

1. Disable ASTScopes for these activities and fall back to the declaration-context-based lookup system. This change will work for the short term, but will require more thought when we completely replace the old context-based lookups. Or,

2. Change ASTScopes to support lookups into inactive clauses. This change will not be easy because of the way that inactive clauses are currently handled in the AST. In addition, @DougGregor  has expressed the intuition that the compiler should not be in the business of lookups into inactive clauses, but I don't know if he would change his mind for this use case.

So, please take a gander at this PR and let me know what you think.

Thanks!

**Comment by davidungar** [comment]
@nathawes Never mind--I think this fix is too draconian. Let me see what I can come up with.

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@nathawes OK, I've got something that just forestalls the problematic lookups. If it tests acceptably, I think it's good. Check it out and let me know. Tx.

**Comment by davidungar** [comment]
@swift-ci please smoke test osx platform

**Comment by davidungar** [comment]
@swift-ci please clean smoke test os x platform

File: lib/IDE/SyntaxModel.cpp [hunk 1]
```diff
@@ -890,15 +906,18 @@ bool ModelASTWalker::walkToDeclPre(Decl *D) {
...
     SyntaxStructureNode SN;
     setDecl(SN, D);
     SN.Kind = SyntaxStructureKind::Extension;
     SN.Range = charSourceRangeFromSourceRange(SM, ED->getSourceRange());
     SN.BodyRange = innerCharSourceRangeFromSourceRange(SM, ED->getBraces());
     SourceRange NSR = SourceRange();
-    if (auto *repr = ED->getExtendedTypeRepr())
-      NSR = repr->getSourceRange();
+    if (!inInactiveClause || ASTScope::areInactiveIfConfigClausesSupported()) {
+      if (auto *repr = ED->getExtendedTypeRepr())
```
Comment (DougGregor) [inline lib/IDE/SyntaxModel.cpp]: Retrieving the type representation should be purely syntactic; I don't think this change should be needed.

Comment (davidungar) [reply lib/IDE/SyntaxModel.cpp]: The compiler crashes without it. Help me out here: what would be a better fix?

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: lib/IDE/SyntaxModel.cpp (hunk 1 above)
Comment (davidungar) [reply lib/IDE/SyntaxModel.cpp]: @DougGregor Thanks to @nathawes , I realized that I completely misunderstood your comment here. I've removed this change. Thanks for your help and patience.

**Review by nathawes — APPROVED** [review]
LGTM - thanks for fixing it!

**Review by DougGregor — APPROVED** [review]

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please clean smoke test os x platform

**Pull request merged**